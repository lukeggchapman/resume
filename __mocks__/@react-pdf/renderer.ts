import React from 'react'

const renderPromise = Promise.resolve()

module.exports = {
  Document: 'Document',
  Font: {
    register: jest.fn((definition: any) => definition),
  },
  Image: 'Image',
  Link: 'Link',
  Page: 'Page',
  PDFViewer: 'PDFViewer',
  StyleSheet: {
    create: jest.fn((styles: any) => styles),
  },
  Text: 'Text',
  View: 'View',
  render: jest.fn((Component: React.ComponentType, props) => renderPromise),
}
