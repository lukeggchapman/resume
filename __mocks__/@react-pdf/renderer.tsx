/* eslint-disable react/function-component-definition */
import React from 'react'

const renderPromise = Promise.resolve()
const container = (name: string) => {
  const ReactPDFContainerMock: React.FC<{ style?: object }> = ({
    style,
    children,
  }) => {
    return (
      <div data-pdf={name} data-style={style && JSON.stringify(style, null, 2)}>
        {children}
      </div>
    )
  }

  ReactPDFContainerMock.displayName = name
  ReactPDFContainerMock.defaultProps = {
    style: undefined,
  }

  return ReactPDFContainerMock
}

module.exports = {
  Document: container('Document'),
  Font: {
    register: jest.fn((definition: any) => definition),
  },
  Image: container('Image'),
  Link: container('Link'),
  Page: container('Page'),
  PDFViewer: 'PDFViewer',
  StyleSheet: {
    create: jest.fn((styles: any) => styles),
  },
  Text: container('Text'),
  View: container('View'),
  render: jest.fn(() => renderPromise),
}
