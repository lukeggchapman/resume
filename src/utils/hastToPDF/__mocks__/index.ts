export default jest.fn((htmlAst) => htmlAst)
