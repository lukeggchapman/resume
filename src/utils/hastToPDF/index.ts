export { default } from './hastToPDF'
export * from './hastToPDF'
