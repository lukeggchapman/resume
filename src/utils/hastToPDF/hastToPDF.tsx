/* eslint-disable no-use-before-define, @typescript-eslint/no-use-before-define */
/**
 * Crude hast (HTML AST) to React-PDF format JSX transform
 */

import React from 'react'
import { uid } from 'react-uid'
import { View, Text } from '@react-pdf/renderer'
import { HastNodes } from 'mdast-util-to-hast/lib'

import { Root, Text as HastText, Element } from 'hast'
import PDFBulletList from './components/PDFBulletList'

const getBulletList = (element: Element) => {
  return element.children
    .filter(
      (node): node is Element =>
        node.type === 'element' && node.tagName === 'li'
    )
    .map((node) => node.children)
    .map((children) => children.map((node) => hastToPDF(node)))
}

const handlers = {
  root: (root: Root): React.ReactNode => {
    const { children } = root

    if (children.length) {
      return children.map(hastToPDF)
    }

    // empty root has no react-PDF equiv
    return undefined
  },
  text: (text: HastText): React.ReactNode => {
    return <Text key={uid(text)}>{text.value}</Text>
  },
  element: (element: Element): React.ReactNode => {
    const { tagName, children } = element

    switch (tagName) {
      case 'p':
        return <View key={uid(element)}>{children.map(hastToPDF)}</View>
      case 'ul':
        return (
          <PDFBulletList key={uid(element)} list={getBulletList(element)} />
        )
      default:
        throw new Error(`hastToPDF: No transform for html element ${tagName}`)
    }
  },
}

function isHandledType(type: string): type is keyof typeof handlers {
  return Object.keys(handlers).indexOf(type) !== -1
}

function hastToPDF(node: HastNodes | null | undefined): React.ReactNode {
  const type = node && node.type

  if (!type) {
    throw new Error(`hastToPDF: Called with invalid root node`)
  }

  if (!isHandledType(type)) {
    throw new Error(`hastToPDF: Cannot compile unknown node ${type}`)
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return handlers[type](node as any)
}

export default hastToPDF
