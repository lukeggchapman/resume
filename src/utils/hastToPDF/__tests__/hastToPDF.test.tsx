import React from 'react'
import { render } from '@testing-library/react'
import { View, Text } from '@react-pdf/renderer'
import { uid } from 'react-uid'
import { HastNodes } from 'mdast-util-to-hast/lib'

import { Root, Text as HastText } from 'hast'
import BulletListPDF from 'components/PDFBulletList'
import hastToPDF from '../hastToPDF'

const bullet1: HastText = {
  type: 'text',
  value: 'point 1.',
}
const bullet2: HastText = {
  type: 'text',
  value: 'point 2.',
}
const hast: Root = {
  type: 'root',
  children: [
    {
      type: 'element',
      tagName: 'p',
      properties: {},
      children: [
        {
          type: 'text',
          value: 'Some paragraph.',
        },
      ],
    },
    {
      type: 'text',
      value: '\n',
    },
    {
      type: 'element',
      tagName: 'ul',
      properties: {},
      children: [
        {
          type: 'text',
          value: '\n',
        },
        {
          type: 'element',
          tagName: 'li',
          properties: {},
          children: [bullet1],
        },
        {
          type: 'text',
          value: '\n',
        },
        {
          type: 'element',
          tagName: 'li',
          properties: {},
          children: [bullet2],
        },
        {
          type: 'text',
          value: '\n',
        },
      ],
    },
    {
      type: 'text',
      value: '\n',
    },
  ],
  data: {
    quirksMode: false,
  },
}

const expected = [
  <View key={uid(hast.children[0])}>
    <Text hyphenationCallback={(e) => [e]}>Some paragraph.</Text>
  </View>,
  <Text
    hyphenationCallback={(e) => [e]}
    key={uid(hast.children[1])}
  >{`\n`}</Text>,
  <BulletListPDF
    key={uid(hast.children[2])}
    list={[
      [
        <Text hyphenationCallback={(e) => [e]} key={uid(bullet1)}>
          point 1.
        </Text>,
      ],
      [
        <Text hyphenationCallback={(e) => [e]} key={uid(bullet2)}>
          point 2.
        </Text>,
      ],
    ]}
  />,
  <Text
    hyphenationCallback={(e) => [e]}
    key={uid(hast.children[3])}
  >{`\n`}</Text>,
]

describe('hastToPDF', () => {
  it('transforms hast to react-pdf jsx', () => {
    const result = hastToPDF(hast)
    const resultContainer = render(<View>{hastToPDF(hast)}</View>)
    const expectedContainer = render(<View>{expected}</View>)

    expect(result).toMatchDiffSnapshot(expected) // shallow
    expect(resultContainer).toMatchDiffSnapshot(expectedContainer) // render
  })

  it('returns nothing with empty hast', () => {
    const emptyHast: HastNodes = {
      type: 'root',
      children: [],
      data: {
        quirksMode: false,
      },
    }

    expect(hastToPDF(emptyHast)).toEqual(undefined)
  })

  it('adds hyphenationCallback to all Text elements', () => {
    const textHast: HastNodes = {
      type: 'root',
      children: [
        {
          type: 'element',
          tagName: 'p',
          properties: {},
          children: [
            {
              type: 'text',
              value: 'Some paragraph.',
            },
          ],
        },
      ],
      data: {
        quirksMode: false,
      },
    }

    const result = hastToPDF(textHast)
    const hyphenationCallback = (result as any)?.[0]?.props?.children[0]?.props
      ?.hyphenationCallback

    expect(hyphenationCallback).toBeDefined()
    expect(hyphenationCallback('test')).toEqual(['test'])
  })

  it('throws error when called with invalid root object', () => {
    expect(() => hastToPDF({} as any)).toThrowError(
      `hastToPDF: Called with invalid root node`
    )
  })

  it('throws error when called with unsupported node type', () => {
    const type = 'random123'
    expect(() => hastToPDF({ type } as any)).toThrowError(
      `hastToPDF: Cannot compile unknown node ${type}`
    )
  })

  it('throws error when called with unsupported html element', () => {
    const tagName = 'strong'
    const element: HastNodes = {
      type: 'element',
      tagName,
      children: [],
    }

    expect(() => hastToPDF(element)).toThrowError(
      `hastToPDF: No transform for html element ${tagName}`
    )
  })
})
