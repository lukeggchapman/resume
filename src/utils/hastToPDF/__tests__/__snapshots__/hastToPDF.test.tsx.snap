// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hastToPDF transforms hast to react-pdf jsx 1`] = `
Snapshot Diff:
Compared values have no visual difference.
`;

exports[`hastToPDF transforms hast to react-pdf jsx 2`] = `
Snapshot Diff:
Compared values have no visual difference.
`;
